File: codeforces/0448D_-_Multiplication_Table/main.cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long int ll;

int n, m;
ll k;

ll f(ll x) {
  ll count = 0;
  for (int i = 1; i <= m; i++) {
    count += min((ll)n, (x - 1) / i);
  }
  return count;
}

int main() {
  cin >> n >> m;
  cin >> k;

  ll low = 1LL;
  ll high = 1LL * n * m + 1;
  while (low < high) {
    ll mid = low + (high - low) / 2;
    if (f(mid) < k)
      low = mid + 1;
    else
      high = mid;
  }

  cout << low - 1 << "\n";
  
  return 0;
}

File: codeforces/0448D_-_Multiplication_Table/main.py
import sys

def main(n, m, k):
    def f(x, n, m, k):
        return sum([min(n, (x - 1) // i) for i in range(1, m + 1)]) < k
    low = 1
    high = n * m + 1
    while low < high:
        mid = low + (high - low)//2
        if f(mid, n, m, k):
            low = mid + 1
        else:
            high = mid
    return low - 1

sys.stdout.write('{}'.format(main(*list(map(int, sys.stdin.readline().split(' '))))))
